### PR TITLE
Explore: adds Loki explore query editor

### LINKED ISSUE #19544: Explore: With Loki the query input overflows on long query
When query gets long the Loki query input jumps to next line:
Screenshot 2019-10-01 at 18 16 49

There is also another issue open that probably could be tackled together if possible (as I assume by fixing this one the other one will again be valid): https://github.com/grafana/grafana/issues/15246

### PR DESCRIPTION
**Pull request by Estrax**
**What this PR does / why we need it**:

Adds LokiExploreQueryEditor - a wrapper on top of LokiQueryField allowing to change lines limit in logs mode. Also, fixes Loki input overflows on long queries and moves `ExploreMode` to grafana-data.

**Which issue(s) this PR fixes**:

Fixes #19544

### CODE REVIEW

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx [hunk 1]
```diff
@@ -173,8 +182,13 @@ export class LokiQueryFieldForm extends React.PureComponent<LokiQueryFieldFormPr
               syntaxLoaded={syntaxLoaded}
             />
           </div>
+          {children}
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: Nitpick: Mentioned it also in some other PR, I think it would be clearer to use some other prop for this like `extraFieldEl` or something like that as the field is not intuitively a wrapper element so using children seem bit weird here.

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  const [maxLines, setMaxLines] = useState('');
+
+  const { isSyntaxReady, setActiveOption, refreshLabels, ...syntaxProps } = useLokiSyntax(
+    datasource.languageProvider,
+    absolute
+  );
+
+  function onMaxLinesChange(e: React.SyntheticEvent<HTMLInputElement>) {
+    const maxLines = e.currentTarget.value;
+    datasource.maxLines = isNaN(+maxLines) || +maxLines <= 0 ? 1000 : +maxLines;
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: We should not modify datasource here. Datasource is a singleton so this change will be persisted and will break other functionality (for example navigating to other datasource and back in explore or in dashboards).

File: packages/grafana-data/src/types/datasource.ts [hunk 3]
```diff
@@ -358,6 +358,7 @@ export interface ExploreQueryFieldProps<
   history: any[];
   onBlur?: () => void;
   absoluteRange?: AbsoluteTimeRange;
+  exploreMode?: 'Logs' | 'Metrics';
```
Comment (aocenas) [inline packages/grafana-data/src/types/datasource.ts]: We have existing [ExploreMode enum](https://github.com/grafana/grafana/blob/c4c031ef4368a84fb35412d8729e78db4ede23cd/public/app/types/explore.ts#L21) that we should use here as a type.

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        onRunQuery={onRunQuery}
+        history={history}
+        data={data}
+        onLoadOptions={setActiveOption}
+        onLabelsRefresh={refreshLabels}
+        syntaxLoaded={isSyntaxReady}
+        absoluteRange={absolute}
+        {...syntaxProps}
+      >
+        {exploreMode === 'Logs' ? (
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: We should use [enum value](https://github.com/grafana/grafana/blob/c4c031ef4368a84fb35412d8729e78db4ede23cd/public/app/types/explore.ts#L21) here for comparison.

**Review by aocenas — CHANGES_REQUESTED** [review]

**Comment by aocenas** [comment]
There are also issues with the latency indicator as it is shown over the max lines input:
Screenshot 2019-12-16 at 15 48 04

Also probably not related to this directly but seems the latency indicator should probably have reserved space so it does not occlude other text:
Screenshot 2019-12-16 at 15 51 30

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  } else {
+    absolute = {
+      from: Date.now() - 10000,
+      to: Date.now(),
+    };
+  }
+
+  const [maxLines, setMaxLines] = useState('');
+
+  const { isSyntaxReady, setActiveOption, refreshLabels, ...syntaxProps } = useLokiSyntax(
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: @kaydelaney we should probably refactor the props of the loki field because right now these syntax props are required but at the same time they are actually created in `LokiQueryField` and so we call `useLokiSyntax` two times.

**Comment by Estrax** [comment]
> There are also issues with the latency indicator as it is shown over the max lines input:
> Screenshot 2019-12-16 at 15 48 04
> 
> Also probably not related to this directly but seems the latency indicator should probably have reserved space so it does not occlude other text:
> Screenshot 2019-12-16 at 15 51 30

Latency has already been moved to QueryRowActions in a separate PR (#20522).

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
@aocenas added a lot of good feedback. I would like to couple more points/questions, mostly regarding UX:

- If I set `Line Limit` to `0` and run the query, I got as many lines, as I've set up in datasource. Shouldn't I get none? 
image

- If I set `Line Limit` to invalid input -`sdnksd` and run the query, I got as many lines, as I've set up in datasource. Shouldn't I also get message that Line Limit is invalid and therefore we have used the limit that was set in datasource? 

- Is Line Limit applicable for each row, or is it for each datasource? Because based on print screen below, I am not sure, but I think that it always takes the higher number and applies it for each query.  So if it is only 1 number for Datasource, I would change where the `LineLimit` input is and not attach it to each query row. 
image

- If the limit was for datasource as a whole, I would change placeholder from `auto` to  `datasource.maxLines`, so you would have there number from datasource, and it would be clear what `LineLimit` is.

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx (hunk 1 above)
Comment (Estrax) [reply public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: Fixed.

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx [hunk 6]
```diff
@@ -173,8 +185,13 @@ export class LokiQueryFieldForm extends React.PureComponent<LokiQueryFieldFormPr
               syntaxLoaded={syntaxLoaded}
             />
           </div>
+          {ExtraFieldElement ? <ExtraFieldElement {...extraFieldProps} /> : null}
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: Is there a reason to instantiate it here? We do not seem to do any logic here so seems like having `ExtraFieldElement` already as an instantiated element and just show it as `{ExtraFieldElement}` would work.

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    datasource.languageProvider,
+    absolute
+  );
+
+  function onChangeQueryLimit(value: string, override?: boolean) {
+    const { query, onChange, onRunQuery } = props;
+    if (onChange) {
+      const nextQuery = { ...query, maxLines: +value };
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: If value is empty string this will set `maxLines: 0`. Is that a correct behaviour here?

File: public/app/core/logs_model.ts [hunk 8]
```diff
@@ -356,7 +361,7 @@ export function logSeriesToLogsModel(logSeries: DataFrame[]): LogsModel | undefi
   if (limits.length > 0) {
     meta.push({
       label: 'Limit',
-      value: `${limits[0].meta.limit} (${rows.length} returned)`,
+      value: `${limit || limits[0].meta.limit} (${rows.length} returned)`,
```
Comment (aocenas) [inline public/app/core/logs_model.ts]: If we set the limit in the query won't it come back in the `meta` of the dataframe? Seems like that would make sense here and we could use that instead of propagating the limit separately.

**Review by aocenas — COMMENTED** [review]
Good work Lukas. I have some small comments here.

**Review by ivanahuckova — COMMENTED** [review]
Based on my previous UX feedback https://github.com/grafana/grafana/pull/21089#pullrequestreview-333316556:

- UX looks beautiful!  I 🧡the consistency with prometheus editor with `Add query` button and `Step` and `Limit` button. Really good job! 
- Amazing job with having and using Limit for each query row separately. 

**Couple questions based on my previous feedback:**

> If the limit was for datasource as a whole, I would change placeholder from auto to datasource.maxLines, so you would have there number from datasource, and it would be clear what LineLimit is.

Any thoughts on this? 

> If I set `Line Limit` to `0` and run the query, I got as many lines, as I've set up in datasource. Shouldn't I get none? 

Any thoughts on this?

> If I set `Line Limit` to invalid input -`sdnksd` and run the query, I got as many lines, as I've set up in datasource. Shouldn't I also get message that Line Limit is invalid and therefore we have used the limit that was set in datasource?

Not sure how big of the deal this is. Based on our conversation - do we want to create a separate issue for this? Or not?

File: packages/grafana-data/src/types/datasource.ts [hunk 9]
```diff
@@ -359,11 +364,12 @@ export interface ExploreQueryFieldProps<
   history: any[];
   onBlur?: () => void;
   absoluteRange?: AbsoluteTimeRange;
+  exploreMode?: ExploreMode.Logs | ExploreMode.Metrics;
```
[the remark this answers is not in the document]
Comment (Estrax) [reply packages/grafana-data/src/types/datasource.ts]: Good catch. I replaced old strings with enum but missed to change them there, just did a fast replace. Fixed.

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx (hunk 6 above)
Comment (Estrax) [reply public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: You're right. Changed it to ReactNode.

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx (hunk 7 above)
Comment (Estrax) [reply public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: Changed it a little - now it falls back to `auto` aka limit of the datasource.

File: public/app/core/logs_model.ts (hunk 8 above)
Comment (Estrax) [reply public/app/core/logs_model.ts]: Yep, we get it back in `meta` of the dataframe, but data can come back in chunks. Already did a small fix to that.

**Comment by Estrax** [comment]
@ivanahuckova regarding your comments:
1) placeholder is still auto, though the initial value of this input is being set up to the `datasource.maxLines` value.
2) Input & result lines count - changed that. For an empty input (visible placeholder "auto") it's going to return a default number of lines for the datasource. When the input value is NaN or just less than or equal to 0, it returns nothing.
3) Need to consult it with UX people.

@aocenas I am pretty sure I overcomplicated these changes, would be glad if you can look at them.

File: public/app/plugins/datasource/loki/result_transformer.ts [hunk 10]
```diff
@@ -447,7 +447,7 @@ export function processRangeQueryResponse(
   switch (response.data.resultType) {
     case LokiResultType.Stream:
       return of({
-        data: lokiStreamsToDataframes(response.data.result, target, limit, config, reverse),
+        data: lokiStreamsToDataframes(response.data.result.slice(0, limit), target, limit, config, reverse),
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/result_transformer.ts]: @Estrax can I ask what is the reason for this change? Does it have anything to do with the explore query editor limit?  Just curious.

**Review by ivanahuckova — COMMENTED** [review]
From the UX point, I love that you have pre-filled the *Line Limit* with the value set in the datasource. For me, it was suddenly much clearer, what  *Line Limit* means, and I interacted with it much more. I also really like, that when I filled *Line Limit* with 0 or text, I didn't get any results. Again, from the UX point, that is something that I would expect happening. 

Regarding issues:
- When I open Explore with Loki datasource, I don't see the Cheatsheet. I see empty graph. When I switch to **Metrics Mode**, I now see Cheatsheet. When I switch to **Logs Mode**, I now see the Cheatsheet. 

Other than that, everything looks beautiful! 😍 😍 😍

File: public/app/plugins/datasource/loki/result_transformer.ts (hunk 10 above)
Comment (Estrax) [reply public/app/plugins/datasource/loki/result_transformer.ts]: Yes - line limit equal to 0 or -1 is technically correct and some datasources return an entire table or document then, so working on a slice is one of the simplest workarounds to that problem.

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/result_transformer.ts]: Makes sense! Thanks for explanation. 🙏

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      const nextQuery = { ...query, maxLines: preprocessMaxLines(value) };
+      onChange(nextQuery);
+
+      if (override && onRunQuery) {
+        onRunQuery();
+      }
+    }
+  }
+
+  function preprocessMaxLines(value: string): number {
```
Comment (aocenas) [inline public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: Only issue I have is with the fall-backing here. If I input `0`, `-10`, `aaa` into the input I get `No logs found.Scan for older logs` message. Which technically isn't true, user just set the limit to be nonsensical in essence. Can we instead of fall-back to last sensible (> 0) value in case of those values? Or maybe even not run the query in that case? Also maybe for this simple case (we want just positive integer) a `type=number` input would make sense.

**Review by aocenas — COMMENTED** [review]
Sorry for the late review @Estrax. Still have one small thing with the input validation/fall-back but otherwise this looks solid.

File: public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx (hunk 11 above)
Comment (Estrax) [reply public/app/plugins/datasource/loki/components/LokiExploreQueryEditor.tsx]: User gets exactly the same message when they have at least one valid query with valid limit and they just hide results coming from this query row, so it made sense to do it this way with line limit values lower than or equal to 0. Just to avoid sending unnecessary requests, I changed `onReturnKeyDown` handler to run queries only when `maxLines` is positive, also added a type param to our `LokiExtraField` and set it to be a number.
